features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,413 @@
+    {
+      PCL_ERROR ("[pcl::%s::computeFeature] No input dataset containing xyz and/or rgb was given!\n", getClassName ().c_str ());
+      return;
+    }
+
+    // RGB -> HSV conversion requires a PointXYZRGB point cloud
+    pcl::PointCloud < pcl::PointXYZRGB > temp_cloud;
+
+    pcl::copyPointCloud (rgb_cloud, temp_cloud);
+
+    // perform RGB -> HSV conversion
+    pcl::PointCloudXYZRGBtoXYZHSV (temp_cloud, hsv_cloud);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointOutT> void
+  GASDColorEstimation<pcl::PointXYZRGB, PointOutT>::convertCloudColorRGBToHSV (const pcl::PointCloud<pcl::PointXYZRGB> &rgb_cloud,
+                                                                               pcl::PointCloud<pcl::PointXYZHSV> &hsv_cloud)
+  {
+    pcl::PointCloudXYZRGBtoXYZHSV (rgb_cloud, hsv_cloud);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDColorEstimationBase<PointInT, PointOutT>::copyColorHistogramsToOutput (const size_t grid_size,
+                                                                             const size_t hists_size,
+                                                                             std::vector<Eigen::VectorXf> &hists,
+                                                                             PointCloudOut &output,
+                                                                             size_t &pos)
+  {
+    for (size_t i = 0; i < grid_size; ++i)
+    {
+      for (size_t j = 0; j < grid_size; ++j)
+      {
+        for (size_t k = 0; k < grid_size; ++k)
+        {
+          const size_t idx = ( (i + 1) * (grid_size + 2) + (j + 1)) * (grid_size + 2) + (k + 1);
+
+          hists[idx][1] += hists[idx][hists_size + 1];
+          hists[idx][hists_size] += hists[idx][0];
+
+          std::copy (hists[idx].data () + 1, hists[idx].data () + hists_size + 1, output.points[0].histogram + pos);
+          pos += hists_size;
+        }
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDColorEstimationBase<PointInT, PointOutT>::computeFeature (PointCloudOut &output)
+  {
+    // call shape feature computation
+    GASDEstimation<PointInT, PointOutT>::computeFeature (output);
+
+    const size_t color_grid_size = color_half_grid_size_ * 2;
+
+    // each histogram dimension has 2 additional bins, 1 in each boundary, for performing interpolation
+    std::vector < Eigen::VectorXf
+        > color_hists ( (color_grid_size + 2) * (color_grid_size + 2) * (color_grid_size + 2), Eigen::VectorXf::Zero (color_hists_size_ + 2));
+
+    pcl::PointCloud < pcl::PointXYZHSV > color_samples;
+
+    convertCloudColorRGBToHSV (shape_samples_, color_samples);
```

Comment:
What about simply ditching this function and directly computing Hue value inside the loop? Avoids memory allocations, needless copying of XYZ values, and all other hassle.